### ISSUE TITLE
Remember installed verbs for a container

### ISSUE THREAD

**Issue by madoar** [PhoenicisOrg/phoenicis#1824]
Currently Phoenicis doesn't remember which verbs were installed in a container. This contains both the verbs installed via a script and the ones installed via the container tab.

To help the user know the current state of his container(s) it is a good idea to save the installed verbs for each container in a file in the container folder. The content of this file could then be used to:
- skip the installation of already installed verbs when executing a script
- display the already installed verbs in the container tab differently (e.g. disable the checkbox, add an "Already installed" text)

**Comment by plata** [comment]
Any preferred technical solution how to store this? Is there maybe a standard solution for this problem?

**Comment by madoar** [comment]
I don't think that there is a standard solution. I think the easiest and best fitting way in our case is in a plain text file, maybe using the `json` format? @qparis do you have a suggestion?

**Comment by qparis** [comment]
We need a bean to read and write into a global configuration file. The implementation is nor important, it could be a local json file as long as we take care about race conditions

**Comment by plata** [comment]
Global? Wouldn't it be better per container?

**Comment by qparis** [comment]
Yep, same principle. We already have a config file per container

**Comment by plata** [comment]
Just to be sure ;)

**Comment by madoar** [comment]
I would like to implement this soon. I think the issue can be divided in two parts:

- changes in Phoenicis:
   - add two new methods to `Engine`:
      ```java
      void registerVerb(String verbId);
      boolean isVerbRegistered(String verbId);
      ```
   - disable the installation checkbox in the container verbs panel when `isVerbRegistered` returns `true` for the given verb
- changes in the Scripts:
   - implement the two new methods `registerVerb` and `isVerbRegistered` for `Wine`
   - add a `registerVerb` call to every `Verb` implementation in the scripts.

Open issues:
Currently the verbs have no knowledge about their script id which is required to call `registerVerb`. I think it would be useful to inject this information long-term in each script. For now I think it is ok to hardcode the script id into each script.

@ImperatorS79 do you want to help with the necessary script changes?

**Comment by ImperatorS79** [comment]
> disable the installation checkbox in the container verbs panel when isVerbRegistered returns true for the given verb

If a verb installation fails somehow, I would like to be possible to reinstall the verb. Maybe add a v next to the verb to say it is already installed ?

**Comment by madoar** [comment]
I would suggest to call `registerVerb` as the last operation in the verb installation. This automatically means that the value is only set when the installation succeeds.

I think a `v` next to the verb name is more confusing to the user. A possibly approach could be the add a `Reinstall` button next to the script, but I think this should be part of a later PR.

**Comment by madoar** [comment]
I just opened a PR with the Java/Phoenicis part (#2070).

@ImperatorS79 do you want to give the JS part a try?

**Comment by madoar** [comment]
The new engine methods look like:

```java
    /**
     * Registers as verb as being "installed"
     *
     * @param verbId The ID of the verb
     * @param container The container where the verb has been installed
     */
    void registerVerb(String verbId, String container);

    /**
     * Checks whether a given verb has been registered as being "installed"
     *
     * @param verbId The ID of the verb
     * @param container The container where the verb has been installed
     * @return True if the verb has been registered as being "installed", false otherwise
     */
    boolean isVerbRegistered(String verbId, String container);
```

**Comment by ImperatorS79** [comment]
We could add the verb to phoenicis.cfg inside a container: verb: [verb1, verb2]. what the verb ID looks like ? I would like this to be still readable.

**Comment by madoar** [comment]
The verb id is the script id, e.g. for adobeair it is `engines.wine.verbs.adobeair`

**Comment by plata** [comment]
> Currently the verbs have no knowledge about their script id which is required to call registerVerb

I don't think it's a good idea to hard code the script IDs. This issue applies to all scripts. Maybe we could solve it with a `ModuleInfo` implemented in Java which returns information about the current or requested module? Somewhat similar to `AppResource`.

**Comment by madoar** [comment]
We need a universal applicable solution that not only works for the `scriptId` but also for example for the `categoryId` and `applicationId` of the installer scripts.

I think your idea is quite could, let me extend it a bit. We could add an additional input to our module lambda function in:

https://github.com/PhoenicisOrg/phoenicis/blob/4fd516e5d7475cded8a8c19d9a0502c9076fbeb6/phoenicis-scripts/src/main/java/org/phoenicis/scripts/engine/injectors/IncludeInjector.java#L35

We could change this to:

```java
String extendedString = String.format("(module, ModuleInfo) => { %s }", script);
```

`ModuleInfo` could then be either a complete custom object, containing fields for e.g. `scriptId`, or a `Map<String, String>`.

I think for completeness we should also change our installer scripts to be modules.

**Comment by plata** [comment]
> ModuleInfo could then be either a complete custom object, containing fields for e.g. scriptId, or a Map<String, String>

Could you give an example how this would look like in Javascript?

> I think for completeness we should also change our installer scripts to be modules.

Yes.

**Comment by madoar** [comment]
Inside the scripts you could access the module infos then like the following:

```javscript
ModuleInfo.SCRIPT_ID // returns the script id
ModuleInfo.APPLICATION_ID // returns the application id (should only work for installer scripts)
...
```

**Comment by plata** [comment]
see https://github.com/PhoenicisOrg/phoenicis/issues/1242
